gtk-ng: don't use signals to toggle command palette

jcollie opened the pull request.

jcollie committed:
gtk-ng: don't use signals to toggle command palette

jcollie commented:
For some reason I'm getting memory leaks when using the command palette that others have been unable to reproduce so there's maybe something with my particular system that is messed up. So I'm not super confident that I haven't introduced a memory leak with this patch.

mitchellh commented on src/apprt/gtk-ng/class/command_palette.zig:
```diff
@@ -216,6 +217,10 @@ pub const CommandPalette = extern struct {
     fn activated(self: *CommandPalette, pos: c_uint) void {
+        // add a reference to keep ourselves around until we're done
+        _ = self.ref();
+        defer self.unref();
```
We should be able to extract the `cmd` from the list model before calling `close` (but also before signal still) so we don't have to do this, because our `cmd` object is still referenced.

jcollie replied:
Fixed.

mitchellh commented on src/apprt/gtk-ng/class/surface.zig:
```diff
@@ -560,13 +547,7 @@ pub const Surface = extern struct {
-            null,
-        );
-        return true;
+        return self.as(gtk.Widget).activateAction("win.toggle-command-palette", null) != 0;
```
Add a comment here that in the future we'll pass the target surface as a param. No need yet.

mitchellh reviewed (changes requested):
One small change requested, and I want to run it through Valgrind after, but looks good!

jcollie committed:
gtk-ng: don't add extra refs when activating a command in the palette

jcollie committed:
gtk-ng: add TODOs about passing surface that toggled command palette

mitchellh reviewed (changes requested):
I pulled this down, showed the palette, and closed the window, and I got tons of Zig leaks.

1. Open Ghostty
2. Open palette
3. Close palette with `esc`
4. Close window (exit or menubar)

mitchellh commented:
Yeah, it doesn't happen on main for me. So its something in this branch (or a heisenbug). I don't have time to dig in yet.

pluiedev commented:
Just the code by itself looks good (can't test Valgrind stuff yet)

jcollie committed:
gtk-ng: fix memory leaks in command palette

jcollie committed:
gtk-ng: better handling of weak references

The upstream GIR for g_weak_ref_get is incorrect - it does not allow the
returned value to be NULL. This PR pulls in a new version of our GObject
bindings with that patched and improves the safety of dealing with the
command palette weak reference held by the window.

See ianprime0509/zig-gobject#117

jcollie commented:
And I think that I've tracked down the memory leaks!

jcollie committed:
gtk-ng: use WeakRef helper for type safety

mitchellh reviewed (commented):
Going to run it later today to test leaks then we should be good.
